max30102.c

```diff
@@ -0,0 +1,123 @@
+                               MAX30102_REG_INTR_ENABLE_1,
+                               data) != MAX30102_OK)
+    {
+        return MAX30102_ERROR;
+    }
+
+    /* 18-bit ADC resolution (SPO2_ADC_RGE = 3) */
+    /* 50 samples per second (SPO2_SR = 2) */
+    /* 411 μs LED pulse width (LED_PW = 3)  */
+    /* Combined as 0x63 for the SPO2_CONFIG register */
+
+    data = 0x63; //01100011
+    if (MAX30102_WriteRegister(hi2c,
+                               MAX30102_REG_SPO2_CONFIG,
+                               data) != MAX30102_OK)
+    {
+        return MAX30102_ERROR;
+    }
+
+    // initial LED pulse amplitudes
+    data = 0x24;
+    if (MAX30102_WriteRegister(hi2c,
+                               MAX30102_REG_LED1_PA,
+                               data) != MAX30102_OK)
+    {
+        return MAX30102_ERROR;
+    }
+    if (MAX30102_WriteRegister(hi2c,
+                               MAX30102_REG_LED2_PA,
+                               data) != MAX30102_OK)
+    {
+        return MAX30102_ERROR;
+    }
+
+    return MAX30102_OK;
+}
+
+MAX30102_Status_t MAX30102_WriteRegister(I2C_HandleTypeDef *hi2c, uint8_t reg, uint8_t value)
+{
+    //write a single byte to the specified register
+    if (HAL_I2C_Mem_Write(hi2c,
+                          MAX30102_I2C_ADDRESS,
+                          reg,
+                          I2C_MEMADD_SIZE_8BIT,
+                          &value,
+                          1,
+                          HAL_MAX_DELAY) != HAL_OK)
+    {
+        return MAX30102_ERROR;
+    }
+
+    return MAX30102_OK;
+}
+
+MAX30102_Status_t MAX30102_ReadRegister(I2C_HandleTypeDef *hi2c, uint8_t reg, uint8_t *value)
+{
+    //read a single byte from the specified register
+    if (HAL_I2C_Mem_Read(hi2c,
+                         MAX30102_I2C_ADDRESS,
+                         reg,
+                         I2C_MEMADD_SIZE_8BIT,
+                         value,
+                         1,
+                         HAL_MAX_DELAY) != HAL_OK)
+    {
+        return MAX30102_ERROR;
+    }
+
+    return MAX30102_OK;
+}
+
+MAX30102_Status_t MAX30102_ReadFIFO(I2C_HandleTypeDef *hi2c, uint32_t *red_data, uint32_t *ir_data) {
+    uint8_t fifo_data[6]; 
+
+    // Read 6 bytes from the FIFO data register
+    if (HAL_I2C_Mem_Read(hi2c,
+                         MAX30102_I2C_ADDRESS,
+                         MAX30102_REG_FIFO_DATA,
+                         I2C_MEMADD_SIZE_8BIT,
+                         fifo_data,
+                         6,
+                         HAL_MAX_DELAY) != HAL_OK)
+    {
+        return MAX30102_ERROR;
+    }
+    
+    // Intensity of red LED reflected back into MAX30102 photodetector
+    *red_data = ((uint32_t)fifo_data[0] << 16) | ((uint32_t)fifo_data[1] << 8) | fifo_data[2];
+    // Intensity of infrared LED reflected back into MAX30102 photodetector
+    *ir_data = ((uint32_t)fifo_data[3] << 16) | ((uint32_t)fifo_data[4] << 8) | fifo_data[5]; 
+
+    return MAX30102_OK;
+}
```

Comment:
@Mame-Mor-M , explain in detail

Reply:
For both red and ir (infrared) data we are shifting the most significant byte left by 16-bits and storing it into the first 8-bits of red/ir_data. The process is repeated for the middle byte where it is shifted left by 8 bits so that it is stored in the next 8 bits, and finally the least-significant bytes are stored into the last 8-bits. The first 3 bytes represent the red data [FIRST CHANNEL] while the next 3 bytes represent infrared data [SECOND CHANNEL]. In the end a 24-bit integer is created for both red_data and ir_data. 24-bits is required as 3 bytes of data are returned by the MAX30102 sensor for both infrared and red light data as demonstrated in the datasheet.